README.md
# Color Hash

Deterministically generate color based on a given string.

[![Build Status](https://travis-ci.org/hugojosefson/color-hash.svg?branch=master)](https://travis-ci.org/hugojosefson/color-hash)
[![npm page](https://img.shields.io/npm/v/@hugojosefson/color-hash.svg)](https://npmjs.com/package/@hugojosefson/color-hash)
[![License MIT](https://img.shields.io/npm/l/@hugojosefson/color-hash.svg)](https://tldrlegal.com/license/mit-license)
[![SemVer 2.0.0](https://img.shields.io/badge/SemVer-2.0.0-lightgrey.svg)](http://semver.org/spec/v2.0.0.html)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

[![Waffle.io](https://img.shields.io/waffle/label/hugojosefson/color-hash/inbox.svg)](https://waffle.io/hugojosefson/color-hash)
[![Waffle.io](https://img.shields.io/waffle/label/hugojosefson/color-hash/to%20do.svg)](https://waffle.io/hugojosefson/color-hash)
[![Waffle.io](https://img.shields.io/waffle/label/hugojosefson/color-hash/in%20progress.svg)](https://waffle.io/hugojosefson/color-hash)

_Based on [zenozeng/color-hash](https://github.com/zenozeng/color-hash); forked to restructure and future-proof._

## Demo

https://hugojosefson.github.io/color-hash/demo/

## Install

```bash
yarn add @hugojosefson/color-hash

 # or

npm install --save @hugojosefson/color-hash
```

## Usage

```javascript
const ColorHash = require('@hugojosefson/color-hash');
const colorHash = new ColorHash();

// in HSL, Hue ∈ [0, 360), Saturation ∈ [0, 1], Lightness ∈ [0, 1]
colorHash.hsl('Hello World'); // [ 225, 0.65, 0.35 ]

// in RGB, R, G, B ∈ [0, 255]
colorHash.rgb('Hello World'); // [ 135, 150, 197 ]

// in HEX
colorHash.hex('Hello World'); // '#8796c5'
```

### Custom Hash Function

```javascript
const ColorHash = require('@hugojosefson/color-hash');

const customHash = function(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash += str.charCodeAt(i);
    }
    return hash;
};

const colorHash = new ColorHash({hash: customHash});
colorHash.hsl('Hello World!');
colorHash.rgb('Hello World!');
colorHash.hex('Hello World!');
```

### Custom Hue

```javascript
const colorHash = new ColorHash({hue: 90});
```

```javascript
const colorHash = new ColorHash({hue: {min: 90, max: 270}});
```

```javascript
const colorHash = new ColorHash({hue: [ {min: 30, max: 90}, {min: 180, max: 210}, {min: 270, max: 285} ]});
```

### Custom Lightness

```javascript
const colorHash = new ColorHash({lightness: 0.5});
```

```javascript
const colorHash = new ColorHash({lightness: [0.35, 0.5, 0.65]});
```

### Custom Saturation

```javascript
const colorHash = new ColorHash({saturation: 0.5});
```

```javascript
const colorHash = new ColorHash({saturation: [0.35, 0.5, 0.65]});
```

## Dev

### Test

```bash
yarn
yarn test
```

#### Coverage Report

https://hugojosefson.github.io/color-hash/coverage/lcov-report/src/

### Build standalone `dist/color-hash.js`

```bash
yarn
yarn build:dist
```

### Build coverage report and demo

```bash
yarn
yarn build:gh-pages
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

_New unreleased changes are listed here, and moved down to a release number when released._

## [2.0.3] - 2017-10-05

### Added

- `keywords` in `package.json`.

## [2.0.2] - 2017-10-05

### Added

- This file, `CHANGELOG.md`.

## [2.0.1] - 2017-10-04

### Changed

- Restructure build, including directories.

## [2.0.0] - 2017-10-04

### BREAKING

- Using [`string-hash`](https://www.npmjs.com/package/string-hash) by default. This changes color of existing strings. Reason for switch is that `string-hash` gives more useful hashes for our purpose. The previous hash function returned same hash for different strings of same length.

## [1.1.0] - 2017-10-04

### Added

- Correctly calculate hue in a range.
- Support multiple hue ranges.
- Test coverage for hue range(s).
- Automatically generate test coverage and browser build before publishing to npm.
- `yarn.lock`
- Connect to [Travis-CI](https://travis-ci.org/hugojosefson/color-hash).
- Publish to npm as [`@hugojosefson/color-hash`](https://npmjs.com/package/@hugojosefson/color-hash).

### Changed

- No need to require `npm install -g` of tools/dependencies.
- Urls etc for forking and publishing to npm scope.
- Building the lib now requires `node@>=6`.
- Improve build.

### Removed

- `bower.json`

---

## [FORKED] - 2017-10-04

Previous releases from 2015-05-18 and earlier via [zenozeng/color-hash](https://github.com/zenozeng/color-hash).


test/bkdr-hash.js
var BKDRHash = require('../src/bkdr-hash');
var assert = require('assert');

describe('BKDRHash', function() {
    it('should return different value for different string', function() {
        assert.notEqual(BKDRHash('abc'), BKDRHash('hij'));
    });

    it('should work even if the string is very long', function() {
        var longstr = '';
        for(var i = 0; i < 10 * 1000; i++) {
            longstr += "Hello World.";
        };
        var hash = BKDRHash(longstr);
        assert.notEqual(hash, Infinity);
        assert.notEqual(hash, 0);
    });

    it('should return different value for different long string', function() {
        var longstr = '';
        for(var i = 0; i < 10 * 1000; i++) {
            longstr += "Hello World.";
        };
        var hash1 = BKDRHash(longstr);
        var hash2 = BKDRHash(longstr.substring(0, longstr.length - 1) + 'x');
        assert.notEqual(hash1, hash2);
    });
});


test/color-hash.js
/* eslint-env mocha */
var assert = require('assert')
var rewire = require('rewire')
var ColorHash = rewire('../src')
var HSL2RGB = ColorHash.__get__('HSL2RGB')
var RGB2HEX = ColorHash.__get__('RGB2HEX')

function assertHueWithinRange (options, expectedRange) {
  return assertHueWithinRanges(options, [expectedRange])
}

function assertHueWithinRanges (options, ranges) {
  options = options || {}
  if (typeof ranges === 'undefined' || !ranges.length || !ranges[0]) {
    ranges = [{}]
  }

  function hueOf (s) {
    var colorHash = new ColorHash(options)
    var hsl = colorHash.hsl(s)
    return hsl[0]
  }

  ranges = ranges.map(function (range) {
    return {
      min: range.min,
      max: range.max,
      size: range.max - range.min,
      minSeen: 1000,
      maxSeen: -1000
    }
  })

  var totalSize = ranges.reduce(function (sum, range) {
    return sum + range.size
  }, 0)

  var iterations = 10 * (totalSize + 1)
  var hue
  for (var i = 0; i < iterations; i++) {
    hue = hueOf('This is some padding, and then a counter: ' + i)
    assert.ok(hue >= 0 || hue < 360, JSON.stringify(ranges, null, 2) + ' hue=' + hue + ' is outside parameters')
    var withinAtLeastOneRange = ranges.reduce(function (withinAnyRangeYet, range) {
      var withinThisRange = hue >= range.min && hue <= range.max
      if (withinThisRange) {
        range.minSeen = Math.min(range.minSeen, hue)
        range.maxSeen = Math.max(range.maxSeen, hue)
      }
      return withinAnyRangeYet || withinThisRange
    }, false)
    assert.ok(withinAtLeastOneRange, JSON.stringify(ranges, null, 2) + ' hue=' + hue + ' is not within any range.')
  }
  ranges.forEach(function (range) {
    assert.equal(Math.round(range.minSeen), range.min, '{min: ' + range.min + ', max: ' + range.max + '} actual minSeen=' + range.minSeen)
    assert.equal(Math.round(range.maxSeen), range.max, '{min: ' + range.min + ', max: ' + range.max + '} actual maxSeen=' + range.maxSeen)
  })
}

describe('ColorHash', function () {
  describe('#Hue', function () {
    it('should return the hash color based on default hue', function () {
      assertHueWithinRange(undefined, {min: 0, max: 358}) // hash % 359 means maximum 358
    })

    it('should return the hash color based on numeric hue', function () {
      assertHueWithinRange({hue: 10}, {min: 10, max: 10})
    })

    it('should return the hash color based on same min, max', function () {
      var range = {min: 10, max: 10}
      assertHueWithinRange({hue: range}, range)
    })

    it('should return the hash color based on the given hue {min, max}', function () {
      var min, max, range
      for (min = 0; min < 361; min += 60) {
        for (max = min + 1; max < 361; max += 60) {
          range = {min: min, max: max}
          assertHueWithinRange({hue: range}, range)
        }
      }
    })

    it('should have default value for min if only max is set', function () {
      assertHueWithinRange({hue: {max: 10}}, {min: 0, max: 10})
    })

    it('should have default value for max if only min is set', function () {
      assertHueWithinRange({hue: {min: 350}}, {min: 350, max: 360})
    })

    it('should return the hash color based on different hue ranges', function () {
      var ranges = [
                {min: 30, max: 90},
                {min: 180, max: 210},
                {min: 270, max: 285}
      ]
      assertHueWithinRanges({hue: ranges}, ranges)
    })
  })

  describe('#Lightness & Saturation', function () {
    it('should return the hash color based on default lightness and saturation', function () {
      var colorHash = new ColorHash()
      var hsl = colorHash.hsl('')
      assert.ok([0.35, 0.5, 0.65].includes(hsl[1]))
      assert.ok([0.35, 0.5, 0.65].includes(hsl[2]))
    })

    it('should return the hash color based on the given lightness and saturation', function () {
      var colorHash = new ColorHash({lightness: 0.5, saturation: 0.5})
      var hsl = colorHash.hsl('')
      assert.deepEqual([hsl[1], hsl[2]], [0.5, 0.5])
    })

    it('should return the hash color based on the given lightness array and saturation array', function () {
      var colorHash = new ColorHash({
        lightness: [0.9, 1],
        saturation: [0.9, 1]
      })
      var hsl = colorHash.hsl('')
      assert.ok([0.9, 1].includes(hsl[1]))
      assert.ok([0.9, 1].includes(hsl[2]))
    })
  })

  describe('#CustomHash', function () {
    var customHash = function (str) {
      var hash = 0
      for (var i = 0; i < str.length; i++) {
        hash += str.charCodeAt(i)
      }
      return hash
    }
    var colorHash = new ColorHash({hash: customHash})
    var hsl = [customHash('abc') % 359, 0.35, 0.35]
    var rgb = HSL2RGB(hsl[0], hsl[1], hsl[2])
    var hex = RGB2HEX(rgb)

    it('#hsl: should return the hsl color based on the given hash function', function () {
      assert.deepEqual(colorHash.hsl('abc'), hsl)
    })
    it('#rgb: should return the rgb color based on the given hash function', function () {
      assert.deepEqual(colorHash.rgb('abc'), rgb)
    })
    it('#hex: should return the hex color based on the given hash function', function () {
      assert.deepEqual(colorHash.hex('abc'), hex)
    })
  })
})


test/string-hash.js
/* eslint-env mocha */
var stringHash = require('string-hash')
var assert = require('assert')

describe('string-hash', function () {
  it('should return different value for different string', function () {
    assert.notEqual(stringHash('abc'), stringHash('hij'))
  })

  it('should work even if the string is very long', function () {
    var longstr = ''
    for (var i = 0; i < 10 * 1000; i++) {
      longstr += 'Hello World.'
    };
    var hash = stringHash(longstr)
    assert.notEqual(hash, Infinity)
    assert.notEqual(hash, 0)
  })

  it('should return different value for different long string', function () {
    var longstr = ''
    for (var i = 0; i < 10 * 1000; i++) {
      longstr += 'Hello World.'
    };
    var hash1 = stringHash(longstr)
    var hash2 = stringHash(longstr.substring(0, longstr.length - 1) + 'x')
    assert.notEqual(hash1, hash2)
  })
})


src/index.js
var stringHash = require('string-hash')

/**
 * Convert RGB Array to HEX
 *
 * @param {Array} RGBArray - [R, G, B]
 * @returns {String} 6 digits hex starting with #
 */
var RGB2HEX = function (RGBArray) {
  var hex = '#'
  RGBArray.forEach(function (value) {
    if (value < 16) {
      hex += 0
    }
    hex += value.toString(16)
  })
  return hex
}

/**
 * Convert HSL to RGB
 *
 * @see {@link http://zh.wikipedia.org/wiki/HSL和HSV色彩空间} for further information.
 * @param {Number} H Hue ∈ [0, 360)
 * @param {Number} S Saturation ∈ [0, 1]
 * @param {Number} L Lightness ∈ [0, 1]
 * @returns {Array} R, G, B ∈ [0, 255]
 */
var HSL2RGB = function (H, S, L) {
  H /= 360

  var q = L < 0.5 ? L * (1 + S) : L + S - L * S
  var p = 2 * L - q

  return [H + 1 / 3, H, H - 1 / 3].map(function (color) {
    if (color < 0) {
      color++
    }
    if (color > 1) {
      color--
    }
    if (color < 1 / 6) {
      color = p + (q - p) * 6 * color
    } else if (color < 0.5) {
      color = q
    } else if (color < 2 / 3) {
      color = p + (q - p) * 6 * (2 / 3 - color)
    } else {
      color = p
    }
    return Math.round(color * 255)
  })
}

function isArray (o) {
  return Object.prototype.toString.call(o) === '[object Array]'
}

/**
 * Color Hash Class
 *
 * @class
 */
var ColorHash = function (options) {
  options = options || {}

  var LS = [options.lightness, options.saturation].map(function (param) {
    param = param || [0.35, 0.5, 0.65] // note that 3 is a prime
    return isArray(param) ? param.concat() : [param]
  })

  this.L = LS[0]
  this.S = LS[1]

  if (typeof options.hue === 'number') {
    options.hue = {min: options.hue, max: options.hue}
  }
  if (typeof options.hue === 'object' && !isArray(options.hue)) {
    options.hue = [options.hue]
  }
  if (typeof options.hue === 'undefined') {
    options.hue = []
  }
  this.hueRanges = options.hue.map(function (range) {
    return {
      min: typeof range.min === 'undefined' ? 0 : range.min,
      max: typeof range.max === 'undefined' ? 360 : range.max
    }
  })

  this.hash = options.hash || stringHash
}

/**
 * Returns the hash in [h, s, l].
 * Note that H ∈ [0, 360); S ∈ [0, 1]; L ∈ [0, 1];
 *
 * @param {String} str string to hash
 * @returns {Array} [h, s, l]
 */
ColorHash.prototype.hsl = function (str) {
  var H, S, L
  var hash = this.hash(str)

  if (this.hueRanges.length) {
    var range = this.hueRanges[hash % this.hueRanges.length]
    var hueResolution = 727 // note that 727 is a prime
    H = ((hash / this.hueRanges.length) % hueResolution) * (range.max - range.min) / hueResolution + range.min
  } else {
    H = hash % 359 // note that 359 is a prime
  }
  hash = parseInt(hash / 360)
  S = this.S[hash % this.S.length]
  hash = parseInt(hash / this.S.length)
  L = this.L[hash % this.L.length]

  return [H, S, L]
}

/**
 * Returns the hash in [r, g, b].
 * Note that R, G, B ∈ [0, 255]
 *
 * @param {String} str string to hash
 * @returns {Array} [r, g, b]
 */
ColorHash.prototype.rgb = function (str) {
  var hsl = this.hsl(str)
  return HSL2RGB.apply(this, hsl)
}

/**
 * Returns the hash in hex
 *
 * @param {String} str string to hash
 * @returns {String} hex with #
 */
ColorHash.prototype.hex = function (str) {
  var rgb = this.rgb(str)
  return RGB2HEX(rgb)
}

module.exports = ColorHash


demo/color-hash.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.ColorHash = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";

function hash(str) {
  var hash = 5381,
      i    = str.length;

  while(i) {
    hash = (hash * 33) ^ str.charCodeAt(--i);
  }

  /* JavaScript does bitwise operations (like XOR, above) on 32-bit signed
   * integers. Since we want the results to be always positive, convert the
   * signed int to an unsigned by doing an unsigned bitshift. */
  return hash >>> 0;
}

module.exports = hash;

},{}],2:[function(require,module,exports){
var stringHash = require('string-hash')

/**
 * Convert RGB Array to HEX
 *
 * @param {Array} RGBArray - [R, G, B]
 * @returns {String} 6 digits hex starting with #
 */
var RGB2HEX = function (RGBArray) {
  var hex = '#'
  RGBArray.forEach(function (value) {
    if (value < 16) {
      hex += 0
    }
    hex += value.toString(16)
  })
  return hex
}

/**
 * Convert HSL to RGB
 *
 * @see {@link http://zh.wikipedia.org/wiki/HSL和HSV色彩空间} for further information.
 * @param {Number} H Hue ∈ [0, 360)
 * @param {Number} S Saturation ∈ [0, 1]
 * @param {Number} L Lightness ∈ [0, 1]
 * @returns {Array} R, G, B ∈ [0, 255]
 */
var HSL2RGB = function (H, S, L) {
  H /= 360

  var q = L < 0.5 ? L * (1 + S) : L + S - L * S
  var p = 2 * L - q

  return [H + 1 / 3, H, H - 1 / 3].map(function (color) {
    if (color < 0) {
      color++
    }
    if (color > 1) {
      color--
    }
    if (color < 1 / 6) {
      color = p + (q - p) * 6 * color
    } else if (color < 0.5) {
      color = q
    } else if (color < 2 / 3) {
      color = p + (q - p) * 6 * (2 / 3 - color)
    } else {
      color = p
    }
    return Math.round(color * 255)
  })
}

function isArray (o) {
  return Object.prototype.toString.call(o) === '[object Array]'
}

/**
 * Color Hash Class
 *
 * @class
 */
var ColorHash = function (options) {
  options = options || {}

  var LS = [options.lightness, options.saturation].map(function (param) {
    param = param || [0.35, 0.5, 0.65] // note that 3 is a prime
    return isArray(param) ? param.concat() : [param]
  })

  this.L = LS[0]
  this.S = LS[1]

  if (typeof options.hue === 'number') {
    options.hue = {min: options.hue, max: options.hue}
  }
  if (typeof options.hue === 'object' && !isArray(options.hue)) {
    options.hue = [options.hue]
  }
  if (typeof options.hue === 'undefined') {
    options.hue = []
  }
  this.hueRanges = options.hue.map(function (range) {
    return {
      min: typeof range.min === 'undefined' ? 0 : range.min,
      max: typeof range.max === 'undefined' ? 360 : range.max
    }
  })

  this.hash = options.hash || stringHash
}

/**
 * Returns the hash in [h, s, l].
 * Note that H ∈ [0, 360); S ∈ [0, 1]; L ∈ [0, 1];
 *
 * @param {String} str string to hash
 * @returns {Array} [h, s, l]
 */
ColorHash.prototype.hsl = function (str) {
  var H, S, L
  var hash = this.hash(str)

  if (this.hueRanges.length) {
    var range = this.hueRanges[hash % this.hueRanges.length]
    var hueResolution = 727 // note that 727 is a prime
    H = ((hash / this.hueRanges.length) % hueResolution) * (range.max - range.min) / hueResolution + range.min
  } else {
    H = hash % 359 // note that 359 is a prime
  }
  hash = parseInt(hash / 360)
  S = this.S[hash % this.S.length]
  hash = parseInt(hash / this.S.length)
  L = this.L[hash % this.L.length]

  return [H, S, L]
}

/**
 * Returns the hash in [r, g, b].
 * Note that R, G, B ∈ [0, 255]
 *
 * @param {String} str string to hash
 * @returns {Array} [r, g, b]
 */
ColorHash.prototype.rgb = function (str) {
  var hsl = this.hsl(str)
  return HSL2RGB.apply(this, hsl)
}

/**
 * Returns the hash in hex
 *
 * @param {String} str string to hash
 * @returns {String} hex with #
 */
ColorHash.prototype.hex = function (str) {
  var rgb = this.rgb(str)
  return RGB2HEX(rgb)
}

module.exports = ColorHash

},{"string-hash":1}]},{},[2])(2)
});